### PR TITLE
Db Integrations scenario: add agent interface validations

### PR DESCRIPTION
**Pull request by robertomonteromiguel**
## Description

We have recently added integration tests for databases.
Until now we have done the validations of the library/trace, in this PR we add the validations of the messages that go out from the agent.

We have also added sql query obfuscation tests.

## Workflow

1. ⚠️⚠️ Create your PR as draft
2. Follow the style guidelines of this project (See [how to easily lint the code](https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md))
3. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
4. Mark it as ready for review

Once your PR is reviewed, you can merge it! :heart:

## Reviewer checklist

* [ ] Check what scenarios are modified. If needed, add the relevant label (`run-parametric-scenario`, `run-profiling-scenario`...). If this PR modifies any system-tests internal, then add the `run-all-scenarios` label ([more info](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md)). 
* [ ] CI is green
   * [ ] If not, failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] if any of `build-some-image` label is present
  1. is the image labl have been updated ? 
  2. just before merging, locally build and push the image to hub.docker.com
* [ ] if a scenario is added (or removed), add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

File: tests/integrations/test_db_integrations_sql.py [hunk 1]
```diff
@@ -192,9 +210,26 @@ def test_error_stack(self):
         span = self._get_sql_span_for_request(self.requests[self.db_service]["select_error"])
         assert span["meta"]["error.stack"].strip()
 
+    def _get_sql_span_for_request(self, weblog_request):
+        """To be implemented by subclasses"""
+        return {}
+
+
+@scenarios.integrations
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:218]: for the code clarety, I think it's better to let those decorator only on actual test classes, rahter than abstract class, WDYT?

File: tests/integrations/test_db_integrations_sql.py [hunk 2]
```diff
@@ -150,7 +168,7 @@ def test_db_row__count(self):
         """ The number of rows/results from the query or operation. For caches and other datastores. 
         This tag should only set for operations that retrieve stored data, such as GET operations and queries, excluding SET and other commands not returning data.  """
         span = self._get_sql_span_for_request(self.requests[self.db_service]["select"])
-        assert span["meta"]["db.row_count"] > 0, f"Test is failing for {db_operation}"
+        assert span["meta"]["db.row_count"] > 0, "Test is failing for select"
```
[the remark this answers is not in the document]
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:171]: I guess for this:

`    @missing_feature(library="python", reason="not implemented yet")
    @missing_feature(library="nodejs", reason="not implemented yet")
    @missing_feature(library="java", reason="not implemented yet")`

We are not executing this test

File: tests/integrations/test_db_integrations_sql.py [hunk 3]
```diff
@@ -192,9 +210,26 @@ def test_error_stack(self):
...
+    def test_NOT_obfuscate_query(self):
+        """ All queries come out without obfuscation from tracer library """
+        for db_operation, request in self.requests[self.db_service].items():
+            span = self._get_sql_span_for_request(request)
+            assert span["resource"].count("?") == 0
+
     def _get_sql_span_for_request(self, weblog_request):
         for data, trace, span in interfaces.library.get_spans(weblog_request):
-            logger.info(f"Span found with trace id: {span['trace_id']} and span id: {span['span_id']}")
+            logger.info(f"Span found with trace id: {span['trace_id']} and span id: {span['span_id']} ")
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:232]: extra space is a typo ?

File: tests/integrations/test_db_integrations_sql.py [hunk 4]
```diff
@@ -192,9 +210,26 @@ def test_error_stack(self):
...
+class _BaseTracerIntegrationsSqlTestClass(_BaseIntegrationsSqlTestClass):
+    @missing_feature(
+        library="java",
+        reason="The Java tracer normalizing the SQL by replacing literals to reduce resource-name cardinality",
+    )
+    def test_NOT_obfuscate_query(self):
+        """ All queries come out without obfuscation from tracer library """
+        for db_operation, request in self.requests[self.db_service].items():
+            span = self._get_sql_span_for_request(request)
+            assert span["resource"].count("?") == 0
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:228]: Maybe a assertion comment explaining the why . And also, `"?" not in span["resource"]` is probably easier to read for non python engineer.

File: tests/integrations/test_db_integrations_sql.py [hunk 5]
```diff
@@ -206,12 +241,68 @@ def _get_sql_span_for_request(self, weblog_request):
                     ):
                         logger.debug("Span type sql found!")
                         logger.info(
-                            f"CHILD Span found with trace id: {span_child['trace_id']} and span id: {span_child['span_id']}"
+                            f"CHILD Span found with trace id: {span_child['trace_id']} and span id: {span_child['span_id']} "
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:244]: extra space is a typo ?

File: tests/integrations/test_db_integrations_sql.py [hunk 6]
```diff
@@ -206,12 +241,68 @@ def _get_sql_span_for_request(self, weblog_request):
...
+                assert db_operation in span["meta"]["sql.query"].lower()
+
+    def test_obfuscate_query(self):
+        """ All queries come out obfuscated from agent """
+        for db_operation, request in self.requests[self.db_service].items():
+            span = self._get_sql_span_for_request(request)
+            # We launch all queries with two parameters (from weblog)
+            # Insert and procedure:These operations also receive two parameters, but are obfuscated as only one.
+            if db_operation in ["insert", "procedure"]:
+                assert span["meta"]["sql.query"].count("?") == 1
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:265]: Assertion comment may help a lot

File: tests/integrations/test_db_integrations_sql.py [hunk 7]
```diff
@@ -206,12 +241,68 @@ def _get_sql_span_for_request(self, weblog_request):
...
+            else:
+                assert span["meta"]["sql.query"].count("?") == 2
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:267]: Assertion comment may help a lot

File: tests/integrations/test_db_integrations_sql.py [hunk 8]
```diff
@@ -206,12 +241,68 @@ def _get_sql_span_for_request(self, weblog_request):
...
+                            return span_child
+
+
+############################################################
+# Postgres: Tracer and Agent validations
+############################################################
+class Test_Tracer_Postgres_db_integration(_BaseTracerIntegrationsSqlTestClass):
+    db_service = "postgresql"
+
+    @missing_feature(library="python", reason="Python is using the correct span: db.system")
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:299]: if all tests are failing on a tracer, it's better to store this info in manifest file. (Same comment for other class in the same case)

File: tests/integrations/test_db_integrations_sql.py [hunk 9]
```diff
@@ -233,7 +327,52 @@ def test_db_user(self):
...
+
+    @missing_feature(library="python", reason="Not implemented yet")
+    @missing_feature(library="java", reason="Not implemented yet")
+    @missing_feature(library="nodejs", reason="Not implemented yet")
+    def test_db_mssql_instance__name(self):
+        """ The Microsoft SQL Server instance name connecting to. This name is used to determine the port of a named instance. 
+            This value should be set only if it’s specified on the mssql connection string. """
+        for db_operation, request in self.requests[self.db_service].items():
+            span = self._get_sql_span_for_request(request)
+            assert span["meta"]["db.mssql.instance_name"].strip(), f"Test is failing for {db_operation}"
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:357]: Assertion comment could be more precise : "`db.mssql.instance_name must not be empty`"

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: tests/integrations/test_db_integrations_sql.py (hunk 1 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:218]: I completely agree

File: tests/integrations/test_db_integrations_sql.py (hunk 4 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:228]: done!

File: tests/integrations/test_db_integrations_sql.py (hunk 3 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:232]: fixed

File: tests/integrations/test_db_integrations_sql.py (hunk 6 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:265]: done!

File: tests/integrations/test_db_integrations_sql.py (hunk 7 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:267]: done!

File: tests/integrations/test_db_integrations_sql.py (hunk 9 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:357]: Done!

File: tests/integrations/test_db_integrations_sql.py (hunk 8 above)
Comment (robertomonteromiguel) [reply tests/integrations/test_db_integrations_sql.py:299]: I added in the manifest for dotnet, golang, php....

File: tests/integrations/test_db_integrations_sql.py [hunk 10]
```diff
@@ -260,3 +409,23 @@ def test_db_system(self):
...
     def test_db_user(self):
         super().test_db_user()
+
+    def test_obfuscate_query(self):
+        """ All queries come out obfuscated from agent """
+        for db_operation, request in self.requests[self.db_service].items():
+            span = self._get_sql_span_for_request(request)
+            # We launch all queries with two parameters (from weblog)
+            # Insert and procedure:These operations also receive two parameters, but are obfuscated as only one.
+            if db_operation in ["insert"] or (db_operation in ["procedure"] and context.library.library != "nodejs"):
```
Comment (cbeauchesne) [inline tests/integrations/test_db_integrations_sql.py:419]: This if/else sequence is a little bit hard to read, and we don't understand why it's different on node. What about this : 

```python
if db_operation == "insert":
   expected_obfuscation_count = 1
elif db_operation == "procedure":
   # explanation why we expect only 0 obfuscation in node
   expected_obfuscation_count = 0 context.library.library == "nodejs" else 1
else:
  expected_obfuscation_count = 2

observed_obfuscation_count = span["meta"]["sql.query"].count("?")
assert (
    observed_obfuscation_count == expected_obfuscation_count
    ), f"The mssql query is not properly obfuscated for operation {db_operation}, expecting {expected_obfuscation_count} obfuscation(s), found {observed_obfuscation_count}:\n {span['meta']['sql.query']}"
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**